tensorflow/dtensor/mlir/utils/collective_lowering.cc

```diff
@@ -285,6 +307,95 @@ mlir::Operation* EmitTransposeOp(mlir::OpBuilder& builder,
                                                perm_op);
 }
 
+mlir::Operation* EmitCollectiveReduceScatter(
+    mlir::OpBuilder& builder, const mlir::Location& loc, mlir::Value input,
+    mlir::Type output_type, const std::string& reduce_op_str,
+    const mlir::DenseIntElementsAttr& group_assignment, int32 scatter_dimension,
+    int32 key_base, mlir::Value device_id, int32 host_group_size,
+    const mlir::StringRef device_type) {
+  mlir::TensorType input_type = input.getType().dyn_cast<mlir::TensorType>();
+
+  const bool need_int32_to_int64_upcast =
+      (device_type.endswith("GPU") && input_type &&
+       input_type.getElementType().isInteger(32));
+
+  const bool need_transpose = scatter_dimension != 0;
+
+  std::vector<int64> perm_for_transpose;
+  if (need_transpose) {
+    perm_for_transpose.reserve(input_type.getRank());
+    for (int i = 0; i < input_type.getRank(); i++) {
+      perm_for_transpose.push_back(i);
+    }
+    std::swap(perm_for_transpose[scatter_dimension], perm_for_transpose[0]);
+    auto pre_transpose_op =
+        EmitTransposeOp(builder, loc, input, perm_for_transpose);
+    input = pre_transpose_op->getResult(0);
+    input_type = input.getType().dyn_cast<mlir::TensorType>();
+    // Compute transposed output type for CollectiveReduceScatter
+    auto output_shape = output_type.dyn_cast<mlir::TensorType>().getShape();
+    std::vector<int64> transposed_shape(output_shape.begin(),
+                                        output_shape.end());
+    for (int i = 0; i < output_shape.size(); i++) {
+      transposed_shape[i] = output_shape[perm_for_transpose[i]];
+    }
+    output_type = mlir::RankedTensorType::get(
+        transposed_shape, need_int32_to_int64_upcast
+                              ? builder.getIntegerType(64)
+                              : input_type.getElementType());
+  }
+
+  if (need_int32_to_int64_upcast) {
+    LOG(WARNING) << "On GPU, collective reduce of int32 is not supported. "
```

Comment:
To confirm, collective scatter reduce also requires an upcast? If yes, can you update the text to reflect reducescatter collective?

Reply:
To be honest, I didn't try running it with int32. There is nothing in NCCL that would prevent it from working, but I know that TensorFlow itself has many limitations for int32 + GPU and I saw similar restrictions throughout DTensor. I was going to ask if anyone working on DTensor knew more about where this upcast came from in `LowerAllReduce`. @rainwoodman 

I can give it a try and see what happens as well.

Reply:
I don't know if the NCCL path CollectiveReduceV2 requires the upcast workaround. When I added the workaround dtensor didn't yet work with NCCL. I'd also suggest try to run without the upcast on the NCCL path, and see what happens.

TensorFlow's distributed runtime doesn't support int32 reductions. The limitation from what I read, indeed runs pretty deep into TensorFlow baked in the assumption to use int32 as a proxy represent host-side shape Tensors for GPU, since the very early days. 

This reminded me we have some NCCL specific unit tests that are not yet OSSed. (Some use multiprocessing and may be a bit complicated to port).

Reply:
Thanks @rainwoodman.

I did some experiments, it looks like both CollectiveReduceV2 and CollectiveReduceScatterV2 work with int32 on GPU using the NCCL path. If it uses the RingReduce path, it will segfault during `tensorflow::collective_util::ComputeBinOp`.

I'll go ahead and remove the upcast since we are always using NCCL for reduce scatter. It looks like we could remove it for CollectiveReduceV2 too if we are using NCCL in a follow-up perhaps.